# Final Write Up

### Goals

#### My goals with this project were to get familiarity with as much of the sklearn library as possible while fulfilling the assignment. 

#### My approach to building a srong predictor is not so much intuitive as it is thourough. I don't know which ones will work best. So  I will try as many as I can.  Given these tools it takes me less time to write several for loops and try different combinations than it does 

### If you put this in edit mode you will see a list of shortcuts. 

Enter           enter edit mode                 Shift-­Enter     run cell, select below
Ctrl-Enter      run cell                        Alt-Enter       run cell, insert below
Y               to code                         M               to markdown
R               to raw                          1               to heading 1
2,3,4,5,6       to heading 2,3,4,5,6            Up/K            select cell above
Down/J          select cell below               A/B             insert cell above/­below
X               cut selected cell               C               copy selected cell
Shift-V         paste cell above                V               paste cell below
Z               undo last cell deletion         D,D             delete selected cell
Shift-M         merge cell below                Ctrl-S          Save and Checkpoint
L               toggle line numbers             O               toggle output
Shift-O         toggle output scrolling         Esc             close pager
H               show keyboard shortcut help     I,I             interrupt kernel
0,0             restart kernel                  Space           scroll down
Shift-­Space    scroll up                       Shift           ignore

Edit Mode (press Enter to enable)
Tab                 code completion or indent   Shift-Tab           tooltip
Ctrl-]              indent                      Ctrl-[              dedent
Ctrl-A              select all                  Ctrl-Z              undo
Ctrl-S­hift-Z       redo                        Ctrl-Y              redo
Ctrl-Home           go to cell start            Ctrl-Up             go to cell start
Ctrl-End            go to cell end              Ctrl-Down           go to cell end
Ctrl-Left           go one word left            Ctrl-Right          go one word right
Ctrl-B­ack­space    delete word before          Ctrl-D­elete         delete word after
Esc                 command mode                Ctrl-M              command mode
Shift-­Enter        run cell, select below      Ctrl-Enter          run cell
Alt-Enter           run cell, insert below      Ctrl-S­hif­t-S­ubtract split cell
Ctrl-S­hift--       split cell                  Ctrl-S              Save and Checkpoint
Ctrl-/              toggle comment on current or selected lines

In [1]:
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.preprocessing import normalize
from sklearn.preprocessing import Imputer
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import StandardScaler
from sklearn import linear_model
from sklearn.cross_validation import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn import metrics
from sklearn.model_selection import ShuffleSplit
from sklearn import svm
from sklearn.naive_bayes import BernoulliNB
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import MultinomialNB
from sklearn.naive_bayes import ClassifierMixin
from sklearn.metrics import accuracy_score
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import GradientBoostingRegressor

path = '/home/collin/machine_learning/CS6140ML18Fall-smith.colin/midterm/data/'
data_file = 'weatherAUS.csv' 

/home/collin/anaconda3/lib/python3.7/site-packages/sklearn/cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)
/home/collin/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/weight_boosting.py:29: DeprecationWarning: numpy.core.umath_tests is an internal NumPy module and should not be imported. It will be removed in a future NumPy release.
  from numpy.core.umath_tests import inner1d


### Import Data

In [2]:
data = pd.read_csv(path + data_file, delimiter=',')

In [3]:
data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145460 entries, 0 to 145459
Data columns (total 24 columns):
Date             145460 non-null object
Location         145460 non-null object
MinTemp          143975 non-null float64
MaxTemp          144199 non-null float64
Rainfall         142199 non-null float64
Evaporation      82670 non-null float64
Sunshine         75625 non-null float64
WindGustDir      135134 non-null object
WindGustSpeed    135197 non-null float64
WindDir9am       134894 non-null object
WindDir3pm       141232 non-null object
WindSpeed9am     143693 non-null float64
WindSpeed3pm     142398 non-null float64
Humidity9am      142806 non-null float64
Humidity3pm      140953 non-null float64
Pressure9am      130395 non-null float64
Pressure3pm      130432 non-null float64
Cloud9am         89572 non-null float64
Cloud3pm         86102 non-null float64
Temp9am          143693 non-null float64
Temp3pm          141851 non-null float64
RainToday        142199 non-null obje

In [4]:
data.describe()

MinTemp        MaxTemp       Rainfall   Evaporation  \
count  143975.000000  144199.000000  142199.000000  82670.000000   
mean       12.194034      23.221348       2.360918      5.468232   
std         6.398495       7.119049       8.478060      4.193704   
min        -8.500000      -4.800000       0.000000      0.000000   
25%         7.600000      17.900000       0.000000      2.600000   
50%        12.000000      22.600000       0.000000      4.800000   
75%        16.900000      28.200000       0.800000      7.400000   
max        33.900000      48.100000     371.000000    145.000000   

           Sunshine  WindGustSpeed   WindSpeed9am   WindSpeed3pm  \
count  75625.000000  135197.000000  143693.000000  142398.000000   
mean       7.611178      40.035230      14.043426      18.662657   
std        3.785483      13.607062       8.915375       8.809800   
min        0.000000       6.000000       0.000000       0.000000   
25%        4.800000      31.000000       7.000000      13.000000   
50%        8.400000      39.000000      13.000000      19.000000   
75%       10.600000      48.000000      19.000000      24.000000   
max       14.500000     135.000000     130.000000      87.000000   

         Humidity9am    Humidity3pm   Pressure9am    Pressure3pm  \
count  142806.000000  140953.000000  130395.00000  130432.000000   
mean       68.880831      51.539116    1017.64994    1015.255889   
std        19.029164      20.795902       7.10653       7.037414   
min         0.000000       0.000000     980.50000     977.100000   
25%        57.000000      37.000000    1012.90000    1010.400000   
50%        70.000000      52.000000    1017.60000    1015.200000   
75%        83.000000      66.000000    1022.40000    1020.000000   
max       100.000000     100.000000    1041.00000    1039.600000   

           Cloud9am      Cloud3pm        Temp9am       Temp3pm        RISK_MM  
count  89572.000000  86102.000000  143693.000000  141851.00000  142193.000000  
mean       4.447461      4.509930      16.990631      21.68339       2.360682  
std        2.887159      2.720357       6.488753       6.93665       8.477969  
min        0.000000      0.000000      -7.200000      -5.40000       0.000000  
25%        1.000000      2.000000      12.300000      16.60000       0.000000  
50%        5.000000      5.000000      16.700000      21.10000       0.000000  
75%        7.000000      7.000000      21.600000      26.40000       0.800000  
max        9.000000      9.000000      40.200000      46.70000     371.000000



# Data Cleaning and manipulation

Because the data in this data set is such a mess it needs to be cleaned up. In order to get the best results I am going to make about a half a dozen datasets and see which is most predictive. There are so many issues with this pile of data. I am going to start with a list of hypotheses. 

1. Wind direction matters, but it will matter "differently" in different locations. The Ocean will be in a different direction in diffrent places. If you just use one dataset my guess is wind direction will look like noise. Separated by loction it will be useful. All of these things are going to be different in different locations. The main dataset will have a predictor. But to the extent it is feasible each location must be dealt with individually. Humidiy variation will have different predictive value inland than it will in coastal areas.

2. This is a dirty dataset. Variations need to be made for each column replacing NaNs with 0, mean, min and max and the combinations run against eachof the variations of hte other columns to be thorough as well as variations that drop each column. The cloumns that are being 1 hot encoded can ignore them. 

3. The dates are going to matter by season. I don't know what the seasons are in australia so month seems like the most likely step. Again twelve new one hot columns minus the date column. This will have to be done before cleaning up sunshine and evaporation. 

4. Sunshine and Evaporation are sensitive to the time of year and by location. But I think more by time of year. The NaN values will be replaced by the average value for that month. It is going to take some effort to find the averages by month for each but the  columns may as well be dropped otherwise. I would like to do this by location as well, but there are 46 of those and 12 months. That would be 552 combinations. =/ 

First step is to turn the date into the month.

Then I will turn the object columns into 1 hot columns with get_dummies.

data_y will be set to RainTomorrow.

Then I will drop the silly columns including the two answer columns. 


In [5]:
date_arr = data['Date']

months = []
for line in date_arr:
    months.append(line[5] + line[6])
    
data_months = pd.Series(months)
data_months.name = 'Month'
    
data = pd.concat([data, data_months], axis=1)

In [6]:
gust_dummies = pd.get_dummies(data['WindGustDir'], prefix='gust_')
nine_am_dummies = pd.get_dummies(data['WindDir9am'], prefix='9am_')
three_pm_dummies = pd.get_dummies(data['WindDir3pm'], prefix='3pm_')
location_dummies = pd.get_dummies(data['Location'])
month_dummies = pd.get_dummies(data['Month'], prefix='month:')
rain_today_dummies = pd.get_dummies(data['RainToday'])

In [7]:
data_with_dummies = pd.concat([data, gust_dummies, nine_am_dummies, three_pm_dummies, location_dummies, month_dummies, rain_today_dummies], axis=1)

data_with_dummies['RainTomorrow'].replace(['No', 'Yes', 'NaN'], [0, 1, 0], inplace=True)

data_y = data_with_dummies["RainTomorrow"]

In [8]:
data_clean = data_with_dummies.drop(columns=
    ['Date', 'Location', 'WindGustDir', 'WindDir9am', 'WindDir3pm', 'Month', 'RainToday', 'RISK_MM'])

cols = data_clean.columns.tolist()

In [9]:
imp = Imputer(missing_values='NaN', strategy='mean', copy=False)

imp.fit(data_clean, data_y)
imputed = imp.transform(data_clean)
data_clean = pd.DataFrame(imputed)

data_clean.columns = cols

## The data above has been dummied and the nan values inputed. 

In [10]:
data_y = data_clean['RainTomorrow']

data_fin = data_clean.drop(columns=['RainTomorrow'])

data_fin_clip = data_fin.clip(lower=0)

for i in range (len(data_y)):
    if data_y[i] < 1:
        data_y[i] = 0

In [11]:
scaler = StandardScaler()
cols = data_fin_clip.columns.tolist()
datax = pd.DataFrame(scaler.fit_transform(data_fin_clip))
datax.columns = cols

In [12]:
l2_data = pd.DataFrame(normalize(data_fin_clip))
l1_data = pd.DataFrame(normalize(data_fin_clip, norm='l1'))
max_data = pd.DataFrame(normalize(data_fin_clip, norm='max'))

l2x_data = pd.DataFrame(normalize(datax))
l1x_data = pd.DataFrame(normalize(datax, norm='l1'))
maxx_data = pd.DataFrame(normalize(datax, norm='max'))

## Setting a baseline. For this I am just checking to see if The data is skewed. It is. 

### Any effective predictor has to improve on .7808

In [13]:
# Set Baseline

yes_arr = rain_today_dummies['Yes']
yes = 0
no = 0
for val in yes_arr:
    if val:
        yes += 1
    else:
        no += 1

if no > yes:
    this = no 
else:
    this = yes

baseline = this / (no + yes)
print('Baseline: ' + str(baseline))

Baseline: 0.7808332187542967


# Data is "Clean"

127  columns dtype: float

data_y is set to 0's and 1's. 

Now we can apply some fancy stuff!

For this dataset I am going to set up a list of training test splits. As I create new test splits for manipulated data sets I will add them to this list. Once the tests snf clssifiers are built ading a new normalised or regularized dataset will be easy to feed into them.

In [14]:
data_fin_clip.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145460 entries, 0 to 145459
Columns: 127 entries, MinTemp to Yes
dtypes: float64(127)
memory usage: 140.9 MB


In [15]:
datax.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145460 entries, 0 to 145459
Columns: 127 entries, MinTemp to Yes
dtypes: float64(127)
memory usage: 140.9 MB


### The blocs below set up  a system by which the train test sets are put in lists and more easily used later on. Each data set has a training and test split which has 4 groups. This allows me to access them as one variable.

In [16]:
t_sets = []
tx_sets = []

TEST_DESC = 0
X_TRAIN = 1
Y_TRAIN = 2
X_TEST = 3
Y_TEST = 4

In [20]:
CLIP = 0

basic=[]
x_train, x_test, y_train, y_test = train_test_split(data_fin_clip, data_y, test_size=.2, random_state=4)
basic.append('Basic DSet: ')
basic.append(x_train)
basic.append(y_train)
basic.append(x_test)
basic.append(y_test)

list=[]
x_train, x_test, y_train, y_test = train_test_split(data_fin_clip, data_y, test_size=.2, random_state=4)
list.append('Clipped Dset: ')
list.append(x_train)
list.append(y_train)
list.append(x_test)
list.append(y_test)
t_sets.append(list)

### Because this is a binary classifier I decided to try the natural binary classifiers in hte Naive-Bayes library.

### Spoiler: They were all terrible.

In [21]:
bern = BernoulliNB()
mult = MultinomialNB()
gaus = GaussianNB()

In [22]:
gaus

GaussianNB(priors=None)

In [23]:
bern_pred = bern.fit(basic[X_TRAIN], basic[Y_TRAIN]).predict(basic[X_TEST])
gaus_pred = gaus.fit(basic[X_TRAIN], basic[Y_TRAIN]).predict(basic[X_TEST])
# mult_pred = mult.fit(t_sets[BAS][X_TRAIN], t_sets[BAS][Y_TRAIN]).predict(t_sets[BAS][X_TEST])


In [24]:
print(accuracy_score(bern_pred, t_sets[0][Y_TEST]))
print(accuracy_score(gaus_pred, t_sets[0][Y_TEST]))


0.7577340849718136
0.6582221916678125


In [25]:
bern_pred = bern.fit(t_sets[CLIP][X_TRAIN], t_sets[CLIP][Y_TRAIN]).predict(t_sets[CLIP][X_TEST])
gaus_pred = gaus.fit(t_sets[CLIP][X_TRAIN], t_sets[CLIP][Y_TRAIN]).predict(t_sets[CLIP][X_TEST])
mult_pred = mult.fit(t_sets[CLIP][X_TRAIN], t_sets[CLIP][Y_TRAIN]).predict(t_sets[CLIP][X_TEST])

In [26]:
print(accuracy_score(bern_pred, t_sets[CLIP][Y_TEST]))
print(accuracy_score(gaus_pred, t_sets[CLIP][Y_TEST]))
print(accuracy_score(mult_pred, t_sets[CLIP][Y_TEST]))


0.7577340849718136
0.6582221916678125
0.7897703836106146


### I prefer to build functions instead of copy pasting. I prefer to put these in a library and import them rather than plugging up my notebook. This works when you have a functioning function.  If you are changing and writing functions as you go they don't update when you save them. Even if you clear everything and re-import it doesn't work. 

### If you are udating a function you are using put it in your notebook. I will keep a library of tested and working functions and add anything new I write to it after I am done with a project.

### append_tts was great here after I closed everything and restarted the computer though. It can be further abstracted.

In [27]:
from funkys_funcs import append_tts

a,b,c,d = train_test_split(l2_data, data_y, test_size=.2, random_state=4)
L2 = append_tts(a,b,c,d, 'L2 Norm Clip:    ', t_sets)

In [28]:
a,b,c,d = train_test_split(l1_data, data_y, test_size=.2, random_state=4)
L1 = append_tts(a,b,c,d, 'L1 Norm Clip:    ', t_sets)

In [29]:
a,b,c,d = train_test_split(max_data, data_y, test_size=.2, random_state=4)
MAX = append_tts(a,b,c,d, 'MAX Norm Clip:   ', t_sets)

In [30]:
a,b,c,d = train_test_split(datax, data_y, test_size=.2, random_state=4)
DATAX = append_tts(a,b,c,d, 'DataX:         ', tx_sets)

In [31]:
a,b,c,d = train_test_split(l2x_data, data_y, test_size=.2, random_state=4)
L2X = append_tts(a,b,c,d, 'L2X Norm Clip:   ', tx_sets)

In [32]:
a,b,c,d = train_test_split(l1x_data, data_y, test_size=.2, random_state=4)
L1X = append_tts(a,b,c,d, 'L1X Norm Clip:   ', tx_sets)

In [42]:
a,b,c,d = train_test_split(maxx_data, data_y, test_size=.2, random_state=4)
MAXX = append_tts(a,b,c,d, 'MAXX Norm:      ', tx_sets)

### I tried the BernoulliNB on base data and all of the normalized options. I actually did this with all three classifiers. Normalizing has no affect on results. Given the poor performance I am not really sure what would improve performance.

In [34]:
bern = BernoulliNB(alpha=2)

for l in tx_sets:
    pred = bern.fit(l[X_TRAIN], l[Y_TRAIN]).predict(l[X_TEST])
    print('Score for ' + l[TEST_DESC] + str(accuracy_score(pred, l[Y_TEST])))

Score for DataX:         0.7806613501993676
Score for L2X Norm Clip:   0.7806613501993676
Score for L1X Norm Clip:   0.7806613501993676
Score for MAXX Norm:      0.7806613501993676


### To look for better results I tried many options. I did the below application with all three classifiers.

In [35]:
for pow in range (-20, 20):
    bern = BernoulliNB(alpha=2**pow)
    print(bern)
    pred = bern.fit(t_sets[1][X_TRAIN], t_sets[1][Y_TRAIN]).predict(l[X_TEST])
    print('Score for ' + l[TEST_DESC] + str(accuracy_score(pred, t_sets[1][Y_TEST])))


BernoulliNB(alpha=9.5367431640625e-07, binarize=0.0, class_prior=None,
      fit_prior=True)
Score for MAXX Norm:      0.7641963426371511
BernoulliNB(alpha=1.9073486328125e-06, binarize=0.0, class_prior=None,
      fit_prior=True)
Score for MAXX Norm:      0.7642650900591228
BernoulliNB(alpha=3.814697265625e-06, binarize=0.0, class_prior=None,
      fit_prior=True)
Score for MAXX Norm:      0.7646775745909529
BernoulliNB(alpha=7.62939453125e-06, binarize=0.0, class_prior=None,
      fit_prior=True)
Score for MAXX Norm:      0.7647119483019387
BernoulliNB(alpha=1.52587890625e-05, binarize=0.0, class_prior=None,
      fit_prior=True)
Score for MAXX Norm:      0.7649525642788395
BernoulliNB(alpha=3.0517578125e-05, binarize=0.0, class_prior=None,
      fit_prior=True)
Score for MAXX Norm:      0.7650213117008112
BernoulliNB(alpha=6.103515625e-05, binarize=0.0, class_prior=None,
      fit_prior=True)
Score for MAXX Norm:      0.7650213117008112
BernoulliNB(alpha=0.0001220703125, binarize=0.

Bernoulli alpha= 2^11
Score for Clipped Dset: 0.7827581465695036
Score for L2 Norm Clip: 0.7827581465695036
Score for L1 Norm Clip: 0.7827581465695036
Score for MAX Norm Clip: 0.7827581465695036

MultinomialNB(alpha=16384, class_prior=None, fit_prior=False)
Score for Clipped Dset: 0.7815206929740135
Score for L2 Norm Clip: 0.7815206929740135
Score for L1 Norm Clip: 0.7815206929740135
Score for MAX Norm Clip: 0.7815206929740135

### With a baseline of .7808 none of these are any good. 

### .7897 was the best result with base multinomial on the most basic clipped dataset and an odd Alpha value

### Discussing the warp and weft of these results will be more useful than me typing up the random thoughts that are driving this.

In [36]:
mult_pred = mult.fit(t_sets[CLIP][X_TRAIN], t_sets[CLIP][Y_TRAIN]).predict(t_sets[CLIP][X_TEST])
print(accuracy_score(mult_pred, t_sets[CLIP][Y_TEST]))


0.7897703836106146


## Gradient Boosting Trees
### GradiantClassifier

In [37]:
gbc = GradientBoostingClassifier(n_estimators=100, learning_rate=1.0, max_depth=1, random_state=0)

pred = gbc.fit(t_sets[1][X_TRAIN], t_sets[1][Y_TRAIN]).predict(t_sets[1][X_TEST])

In [38]:
 print(gbc.score(t_sets[1][X_TEST], t_sets[1][Y_TEST]))

0.8499243778358312


### Now this shows promise. Lots of options and several more datasets to try. 

### First I want to get a single Gradient Boosting Regressor working

In [39]:
grb = GradientBoostingRegressor(n_estimators=100, learning_rate=0.1, max_depth=1, random_state=0, loss='ls')
grb.fit(t_sets[1][X_TRAIN], t_sets[1][Y_TRAIN])

GradientBoostingRegressor(alpha=0.9, criterion='friedman_mse', init=None,
             learning_rate=0.1, loss='ls', max_depth=1, max_features=None,
             max_leaf_nodes=None, min_impurity_decrease=0.0,
             min_impurity_split=None, min_samples_leaf=1,
             min_samples_split=2, min_weight_fraction_leaf=0.0,
             n_estimators=100, presort='auto', random_state=0,
             subsample=1.0, verbose=0, warm_start=False)

### Because this is a regressor it gives me a predicted value

In [40]:
y_pred = grb.predict(t_sets[1][X_TEST])

for i in range(len(y_pred)):
    if y_pred[i] < .5:
        y_pred[i] = 0
    else:
        y_pred[i] = 1

In [43]:
from result_lib import Result

In [44]:
res = Result(y_pred, t_sets[1][Y_TEST].values, .1)

In [45]:
res.display_totals()

Range: 0.1
Correct: 24433
Mean Variance: 0.16014711948301938


In [46]:
24433/len(y_pred)

0.8398528805169806

### .8398 isn't bad, going to try changing the cut off on 1 and 0 from .5

In [47]:
grb = GradientBoostingRegressor(n_estimators=100, learning_rate=0.1, max_depth=1, random_state=0, loss='ls')
y_pred = grb.fit(t_sets[1][X_TRAIN], t_sets[1][Y_TRAIN]).predict(t_sets[1][X_TEST])

for i in range(len(y_pred)):
    if y_pred[i] < .6:
        y_pred[i] = 0
    else:
        y_pred[i] = 1
        
res = Result(y_pred, t_sets[1][Y_TEST].values, .1)

print(res.correct/len(y_pred))

0.8329437646088271


In [48]:
grb = GradientBoostingRegressor(n_estimators=100, learning_rate=0.1, max_depth=1, random_state=0, loss='ls')
y_pred = grb.fit(t_sets[1][X_TRAIN], t_sets[1][Y_TRAIN]).predict(t_sets[1][X_TEST])

for i in range(len(y_pred)):
    if y_pred[i] < .4:
        y_pred[i] = 0
    else:
        y_pred[i] = 1
        
res = Result(y_pred, t_sets[1][Y_TEST].values, .1)

print(res.correct/len(y_pred))

0.8390966588752922


### So far best  result is .8499 whis is significantly above .7808. 

### Now that the basics work the next step would be to try many combinations of estimators and depth and learning rate and regressor cutoffs on the many different variations of the data I currently have.

### My intuition at this point would be to work with the Gradient Boosting regressor. 

#### Given a baseline of .7808 I have a target improvement of 21.92%  and with minimal effort I dropped that to 15.01% The next steps would be to try the combinations in GBC of:

#### alpha=0.9, 
#### criterion='friedman_mse', 
#### init=None,
#### learning_rate=0.1,
#### loss='ls', 
#### max_depth=1, 
#### max_features=None,
#### max_leaf_nodes=None, 
#### min_impurity_decrease=0.0,
#### min_impurity_split=None, 
#### min_samples_leaf=1,
#### min_samples_split=2, 
#### min_weight_fraction_leaf=0.0,
#### n_estimators=100, 
#### presort='auto', 
#### random_state=0,
#### subsample=1.0, 
#### verbose=0, 
#### warm_start=False

### Also Mark Down is not my friend